### PR TITLE
Add value error if both d sae and expansion factor set

### PR DESCRIPTION
**Pull request by anthonyduong9**
# Description

Changes `LanguageModelSAERunnerConfig` so if `d_sae` and `expansion_factor` are not `None`, `__post_init__()` raises a `ValueError` and if `d_sae` and `expansion_factor` are `None`, `__post_init__()` sets `expansion_factor` to `4`. `d_sae` and `expansion_factor` are both attributes to set the size of an SAE, and `expansion_factor` has a default value of `4`, for backward compatibility.

Fixes # https://github.com/jbloomAus/SAELens/issues/47

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### You have tested formatting, typing and unit tests (acceptance tests not currently in use)

- [x] I have run `make check-ci` to check format and linting. (you can run `make format` to format code if needed.)

### Performance Check.

If you have implemented a training change, please indicate precisely how performance changes with respect to the following metrics:
- [ ] L0
- [ ] CE Loss
- [ ] MSE Loss
- [ ] Feature Dashboard Interpretability

Please links to wandb dashboards with a control and test group.

### CODE REVIEW

File: tests/unit/helpers.py [hunk 1]
```diff
@@ -9,51 +9,78 @@
...
-    for key, val in kwargs.items():
-        setattr(mock_config, key, val)
+    for key, value in kwargs.items():
+        mock_config_dict[key] = value
 
-    # Call the post-init method to set any derived attributes
-    # useful for checking the correctness of the configuration
-    # in the tests.
-    mock_config.__post_init__()
+    mock_config = LanguageModelSAERunnerConfig(**mock_config_dict)
```
Comment (chanind) [inline tests/unit/helpers.py:83]: fair enough! This feels less hacky than doing `__post_init__()` like before 👍

File: tests/unit/training/test_activations_store.py [hunk 2]
```diff
@@ -28,7 +28,6 @@ def tokenize_with_bos(model: HookedTransformer, text: str) -> list[int]:
         {
             "model_name": "tiny-stories-1M",
             "dataset_path": "roneneldan/TinyStories",
-            "tokenized": False,
```
Comment (chanind) [inline tests/unit/training/test_activations_store.py:31]: Why were these values changed?

File: tests/unit/training/test_sae_basic.py [hunk 3]
```diff
@@ -202,7 +198,8 @@ def test_sae_save_and_load_from_pretrained_gated(tmp_path: Path) -> None:
 
 def test_sae_save_and_load_from_pretrained_topk(tmp_path: Path) -> None:
     cfg = build_sae_cfg(
-        activation_fn_str="topk", activation_fn_kwargs={"k": 30}, device="cpu"
```
Comment (chanind) [inline tests/unit/training/test_sae_basic.py:205]: why was this changed?

File: tests/unit/training/test_sae_trainer.py [hunk 4]
```diff
@@ -204,7 +204,6 @@ def test_train_sae_group_on_language_model__runs(
     checkpoint_dir = tmp_path / "checkpoint"
     cfg = build_sae_cfg(
         checkpoint_path=str(checkpoint_dir),
-        train_batch_size=32,
```
Comment (chanind) [inline tests/unit/training/test_sae_trainer.py:207]: Why was this changed?

**Review by chanind — COMMENTED** [review]
The core code changes look good, but why were unrelated test values changed?

File: tests/unit/training/test_activations_store.py (hunk 2 above)
Comment (anthonyduong9) [reply tests/unit/training/test_activations_store.py:31]: Because `build_sae_cfg()` now takes a dict, updates another dict, and then passes it to a dataclass, whereas before it instantiated a dataclass, then looped over the dict and called `setattr()` to set the dataclass attributes. So now, since `"tokenized"` doesn't match any attributes of `LanguageModelSAERunnerConfig`, its `__init__()` method would return an error/tests would fail.

I checked throughout the codebase that there are no references to a field `tokenized` for any instance of `LanguageModelSAERunnerConfig`.

File: tests/unit/training/test_sae_basic.py (hunk 3 above)
Comment (anthonyduong9) [reply tests/unit/training/test_sae_basic.py:205]: Similar reasons as in https://github.com/jbloomAus/SAELens/pull/301/files#r1776064470. `activation_fn_str` isn't an attribute of `LanguageModelSAERunnerConfig`.

File: tests/unit/training/test_sae_trainer.py (hunk 4 above)
Comment (anthonyduong9) [reply tests/unit/training/test_sae_trainer.py:207]: For similar reasons as in [this comment](https://github.com/jbloomAus/SAELens/pull/301/files#r1776064470). `train_batch_size` isn't an attribute of LanguageModelSAERunnerConfig.

File: tests/unit/helpers.py (hunk 1 above)
Comment (anthonyduong9) [reply tests/unit/helpers.py:83]: Haha, yeah, we shouldn't effectively call `__post_init__()` twice.

File: tests/unit/training/test_activations_store.py (hunk 2 above)
Comment (chanind) [reply tests/unit/training/test_activations_store.py:31]: oooh nice, so if we mistype attributes in tests now we'll get an error? 🥇

**Comment by chanind** [comment]
Awesome work with this, and thanks for fixing the issues with tests have invalid config fields set!
